domain/resource/modelmigration/export.go

```diff
@@ -0,0 +1,132 @@
+func RegisterExport(
+	coordinator Coordinator,
+	clock clock.Clock,
+	logger logger.Logger,
+) {
+	coordinator.Add(&exportOperation{
+		clock:  clock,
+		logger: logger,
+	})
+}
+
+type ExportService interface {
+	// ExportResources returns the list of application and unit resources to
+	// export for the given application.
+	//
+	// The following error types can be expected to be returned:
+	//   - [resourceerrors.ApplicationNotFound] if the application ID is not an
+	//     existing one.
+	//
+	// If the application exists but doesn't have any resources, no error are
+	// returned, the result just contains an empty list.
+	ExportResources(ctx context.Context, name string) (
+		[]coreresource.Resource,
+		[]coreresource.UnitResources,
+		error,
+	)
+}
+
+// exportOperation describes a way to execute a migration for
+// exporting applications.
+type exportOperation struct {
+	modelmigration.BaseOperation
+
+	service ExportService
+
+	registry corestorage.ModelStorageRegistryGetter
+	clock    clock.Clock
+	logger   logger.Logger
+}
+
+// Name returns the name of this operation.
+func (e *exportOperation) Name() string {
+	return "export resources"
+}
+
+// Setup the export operation.
+// This will create a new service instance.
+func (e *exportOperation) Setup(scope modelmigration.Scope) error {
+	e.service = service.NewService(
+		state.NewState(scope.ModelDB(), e.clock, e.logger),
+		nil,
+		e.logger,
+	)
+	return nil
+}
+
+// Execute the export, adding the application to the model.
+// The export also includes all the charm metadata, manifest, config and
+// actions. Along with units and resources.
+func (e *exportOperation) Execute(ctx context.Context, model description.Model) error {
+	for _, app := range model.Applications() {
+		resources, unitResources, err := e.service.ExportResources(ctx, app.Name())
+		if err != nil {
+			return errors.Errorf("getting resource of application %s: %w", app.Name(), err)
+		}
+
+		for _, res := range resources {
+			r := app.AddResource(description.ResourceArgs{
+				Name: res.Name,
+			})
+			r.SetApplicationRevision(description.ResourceRevisionArgs{
+				Type:        res.Type.String(),
+				Origin:      res.Origin.String(),
+				Timestamp:   res.Timestamp,
+				Revision:    res.Revision,
+				SHA384:      res.Fingerprint.String(),
+				Size:        res.Size,
+				RetrievedBy: res.RetrievedBy,
+			})
+		}
+
+		if len(unitResources) == 0 {
+			continue
+		}
+
+		unitNameToResources := make(map[string][]coreresource.Resource)
+		for _, unitRes := range unitResources {
+			unitNameToResources[unitRes.Name.String()] = unitRes.Resources
+		}
+
+		for _, unit := range app.Units() {
+			unitResources := unitNameToResources[unit.Name()]
+			for _, unitRes := range unitResources {
+				unit.AddResource(description.UnitResourceArgs{
+					Name: unitRes.Name,
+					RevisionArgs: description.ResourceRevisionArgs{
+						Type:        unitRes.Type.String(),
```

Comment:
question: Type can be found in the charm resource data which is migrated. As it's immutable, why copy it here?

Reply:
For Type I would like to keep exporting it, even though we don't import it. In the past we have had a lot of branching logic on the type of the resource. Of course we want to get rid of that, and have done during the work of this epic, but the point remains that it is an import piece of information about the resource. 

For me, I think this is worth exporting "just in case", even if we never use it. The removed `Description` and `Path` fields are far less likely to ever be useful on the import, the first is only used for metadata, and the second is used once the resource gets to the unit, which is necessarily a few steps after migration and requires the resource to have been successfully entered into the database and object store. The `Type` does not posses these properties.